### PR TITLE
SNOW-817445: Fix dataframe writer save_as_table allow nulls

### PR DESCRIPTION
**Pull request by sfc-gh-aalam**
Please answer these questions before submitting your pull requests. Thanks!

1. What GitHub issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   Fixes #SNOW-817445

2. Fill out the following pre-review checklist:

   - [x] I am adding a new automated test(s) to verify correctness of my new code
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency

3. Please describe how your code solves the related issue.

   Schema provided to database is not entirely respected when save as table is called. This change makes sure we respect schema.

### CODE REVIEW

File: src/snowflake/snowpark/_internal/analyzer/snowflake_plan.py [hunk 1]
```diff
@@ -572,57 +604,23 @@ def save_as_table(
...
         elif mode == SaveMode.IGNORE:
-            return self.build(
-                lambda x: create_table_as_select_statement(
-                    full_table_name, x, error=False, table_type=table_type
-                ),
-                child,
-                None,
-            )
+            # return get_create_and_insert_plan(replace=False, error=False)
+            if self.session._table_exists(table_name):
```
Comment (sfc-gh-aalam) [inline src/snowflake/snowpark/_internal/analyzer/snowflake_plan.py:612]:
```suggestion
            if self.session._table_exists(table_name):
```

File: src/snowflake/snowpark/_internal/analyzer/analyzer_utils.py [hunk 2]
```diff
@@ -450,6 +452,27 @@ def range_statement(start: int, end: int, step: int, column_name: str) -> str:
...
 
 
+def schema_query_for_values_statement(output: List[Attribute]) -> str:
+    data_types = [attr.datatype for attr in output]
+    names = [quote_name(attr.name) for attr in output]
+    nullables = [attr.nullable for attr in output]
+    cells = [
+        schema_expression(data_type, is_nullable)
+        for data_type, is_nullable in zip(data_types, nullables)
+    ]
```
Comment (sfc-gh-sfan) [inline src/snowflake/snowpark/_internal/analyzer/analyzer_utils.py:462]: Extracting the info into separate lists and zip them here feels a bit counterintuitive. Should we just do a `for attr in output`?

File: src/snowflake/snowpark/_internal/analyzer/analyzer_utils.py [hunk 3]
```diff
@@ -450,6 +452,27 @@ def range_statement(start: int, end: int, step: int, column_name: str) -> str:
...
+
+    query = (
+        SELECT
+        + COMMA.join([f"{DOLLAR}{i+1}{AS}{c}" for i, c in enumerate(names)])
```
Comment (sfc-gh-sfan) [inline src/snowflake/snowpark/_internal/analyzer/analyzer_utils.py:466]: Extracting into `names` feels unnecessary if the usage is `enumerate`.

File: src/snowflake/snowpark/_internal/analyzer/snowflake_plan.py [hunk 4]
```diff
@@ -560,6 +561,37 @@ def save_as_table(
         child: SnowflakePlan,
     ) -> SnowflakePlan:
         full_table_name = ".".join(table_name)
+
+        def get_create_and_insert_plan(replace=False, error=True):
```
Comment (sfc-gh-sfan) [inline src/snowflake/snowpark/_internal/analyzer/snowflake_plan.py:565]: Nice :)

File: src/snowflake/snowpark/dataframe_writer.py [hunk 5]
```diff
@@ -126,13 +126,13 @@ def save_as_table(
             mode: One of the following values. When it's ``None`` or not provided,
                 the save mode set by :meth:`mode` is used.
 
-                "append": Append data of this DataFrame to existing data.
+                "append": Append data of this DataFrame to the existing table.
```
Comment (sfc-gh-sfan) [inline src/snowflake/snowpark/dataframe_writer.py:129]: Unrelated to your change: I think we should also mention that we will create the table if it does not exist.

**Review by sfc-gh-sfan — APPROVED** [review]

**Pull request merged**